lib/engine/db/postgres/index.ts

```diff
@@ -0,0 +1,150 @@
+//
+
+import * as Tp from 'thingpedia';
+
+import type * as DB from '..';
+
+import { Pool, PoolClient, } from 'pg';
+
+import LocalTable from './local_table';
+import SyncTable from './sync_table';
+
+
+export interface PostgresPlatform extends Tp.BasePlatform {
+    getPostgresConfig() : any;
+}
+
+export function query(client : PoolClient, string : string, args : unknown[]) {
+    return new Promise<void>((resolve, reject) => {
+        client.query(string, args, (err) => {
+            if (err)
+                reject(err);
+            else
+                resolve();
+        });
+    });
+}
+
+export function selectAll(client : PoolClient, string : string, args : unknown[]) {
+    return new Promise<any[]>((resolve, reject) => {
+        client.query(string, args, (err, res) => {
+            if (err)
+                reject(err);
+            else
+                resolve(res.rows);
+        });
+    });
+}
+
+export function selectOne(client : PoolClient, string : string, args : unknown[]) {
+    return new Promise<any>((resolve, reject) => {
+        client.query(string, args, (err, result) => {
+            if (err)
+                reject(err);
+            else
+                resolve(result.rows[0]);
+        });
+    });
+}
+
+export function insertOne(client : PoolClient, string : string, args : unknown[]) {
+    return new Promise((callback, errback) => {
+        const stringReturingId  = string + ' RETURNING uniqueid';
+        client.query(stringReturingId, args, (err, result) => {
+            if (err) {
+                errback(err);
+                return;
+            }
+            if (result === undefined) {
+                errback(new Error('Row does not have uniqueId'));
+                return;
+            }
+            callback(result.rows);
+        });
+    });
+}
+
+const FIELD_NAMES = {
+    app: ['code', 'state', 'name', 'description'] as const,
+    device: ['state'] as const,
+    channel: ['value'] as const
+};
+
+export class PostgresDatabase implements DB.AbstractDatabase {
+    private _pool : Pool;
+    private _transactionQueue = new WeakMap<PoolClient, Promise<unknown>>();
+
+    constructor(public platform : PostgresPlatform) {
+        this._pool = new Pool(platform.getPostgresConfig());
+    }
+
+    withClient<T>(callback : (client : PoolClient) => Promise<T>) : Promise<T> {
+        return new Promise<T>((resolve, reject) => {
+            this._pool.connect((err, c, release) => {
+                if (err) {
+                    console.error('Error in acquiring client', err.stack);
+                    reject(err);
+                    return;
+                }
+                resolve(callback(c));
+                release();
+            });
+        });
+    }
+
+    withTransaction<T>(transaction : (client : PoolClient) => Promise<T>) : Promise<T> {
+        return this.withClient((client) => {
+            const queue = this._transactionQueue.get(client) || Promise.resolve();
```

Comment:
We definitely don't need the transaction queue here, this was a hack to fix broken sqlite's locking. With a real DB, we have real locking and can run multiple transactions concurrently...